salt/_states/kubeconfig.py

```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+
+from base64 import b64encode
+import os
+import stat
+import yaml
+
+__virtualname__ = 'metalk8s_kubeconfig'
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def _validateKubeConfig(filename, expected_ca_data, expected_api_server):
+    """Validate a kubeconfig filename.
+
+    Validate that the kubeconfig provided by filename
+    is conform with config.
+
+    This function is used for managed idempotency.
+
+    :return: True if the kubeconfig file matches expectation
+             False otherwise (ie need to be regenerated)
+    """
+    # Verify that the file does not exist
+    if not os.path.isfile(filename):
+        return False
+
+    # Verify that the mode is 600
+    if stat.S_IMODE(os.stat(filename).st_mode) != 0o600:
+        return False
+
+    try:
+        with open(filename, 'r') as fd:
+            kubeconfig = yaml.safe_load(fd)
+    except Exception:
+        return False
+
+    # Verify that the current CA cert on disk matches the expected CA cert
+    # and the API Server on the existing file match with the expected
+    try:
+        cluster_info = kubeconfig['clusters'][0]['cluster']
+        current_ca_data = cluster_info['certificate-authority-data']
+        current_api_server = cluster_info['server']
+    except (KeyError, IndexError):
+        return False
```

Comment:
You also need to verify `client-certificate-data` because if it's a wrong certificate or certificate expired you need to generate one new but keep the same private key I think.

So for me we need something like :
```
if no private_key: 
  # generate_one
else:
  # get the private key

if certificate:
  # Verify certs field like CN, ...
  # Verify that certs signed by the CA public key (x509.verify_signature)
  # Verify if certs expired or almost expired (x509.will_expire)
  # Verifiy that private key in the certs if the one used in the certificate (x509.verify_private_key, you can use directly the certs as public key)
  if something wrong:
    certificate = None

if not certificate:
  # generate one
```

Reply:
Yep, we can do that but it's really depends on our policy. I did it like kubeadm do and it does not verify the `client-certificate-data`, when you run `kubeadm init  phase kubeconfig kubelet`, and you put a random base64 string in `client-certificate-data`, if you re-run the `kubeadm init  phase kubeconfig kubelet`, it will say that the config if up to date. I will see if the change is not too long, I will do it, otherwise, I will create another ticket

Reply:
Yes because kubeadm is definitly not idempotent. But us we should be.
But btw yes maybe in another PR/tcket

Reply:
In fact, if the file exists, kubeadm check if it is valid, and for his check, it checks only the ca cert, and the API srv.
```
[root@k8s centos]# kubeadm init  phase kubeconfig kubelet
[kubeconfig] Using existing up-to-date kubeconfig file: "/etc/kubernetes/kubelet.conf"
```
I just updated the PR. I took into account all of your remarks except the keep the private key. AFAIK, we don't have a real reason to keep the same priv key as we don't store it. For the cert expiration check, I did not use x509 will expire module because it does not work on pem cert. The docstring says `string containing the PEM formatted text of the certificate.` but it's not implemented in the code